### PR TITLE
Templates for intermediate form generation.

### PR DESCRIPTION
**Pull request by woocash2**

### CODE REVIEW

File: app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package compiler.semantic_analysis
+
+import compiler.ast.Expression
+import compiler.ast.Function
+import compiler.ast.Program
+import compiler.common.diagnostics.Diagnostics
+import compiler.common.semantic_analysis.ReferenceMap
+
+object ArgumentResolver {
+    fun calculateArgumentToParameterResolution(
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt:10]: Can you also add this to Compiler.kt?

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt [hunk 2]
```diff
@@ -38,6 +38,10 @@ sealed class Diagnostic {
         }.toString()
     }
 
+    sealed class ArgumentResolutionErrors() {
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:41]: This should extend Diagnostic. Also should't this be after NameResolutionErrors?

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package compiler.intermediate_form
+
+import compiler.ast.Function
+import compiler.ast.Program
+import compiler.ast.Variable
+import compiler.common.semantic_analysis.ReferenceMap
+
+object FunctionDependenciesAnalyzer {
+    fun variablesUsedByFunctions(ast: Program): ReferenceMap<Function, Pair<List<Variable>, List<Variable>>> { // (accessed, modified)
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt:9]: I think we shouldn't use Pair here, but some simple wrapper data class with named fields instead

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraph.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package compiler.intermediate_form
+
+import compiler.common.semantic_analysis.ReferenceMap
+
+typealias Node = IntermediateFormTreeNode
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraph.kt:5]: Maybe `IFTNode`?

**Review by JHodorHodor — APPROVED** [review]

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt (hunk 2 above)
Comment (tomekb234) [reply app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:41]: Also, the name should be singular `ArgumentResolutionError`.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt (hunk 3 above)
Comment (tomekb234) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt:9]: I would suggest something like `ReferenceMap<Function, ReferenceMap<Variable, VariableAccessMode>>` with `VariableAccessMode` being a two-option enum.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+        return TODO()
+    }
+
+    fun createCallGraph(ast: Program): Set<Pair<Function, Function>> {
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt:13]: A problem here is that we need to compare `Function`s by reference. This should be something like `ReferenceMap<Function, ReferenceSet<Function>>`.

Comment (tomekb234) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt:13]: A `ReferenceSet` can be made in this way:
```kotlin
interface ReferenceSet<E> : Set<E>

interface MutableReferenceSet<E> : MutableSet<E>, ReferenceSet<E>

class ReferenceHashSet<E> : MutableSet<E> by Collections.newSetFromMap(ReferenceHashMap()), MutableReferenceSet<E>
```

File: app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        ast: Program,
+        diagnostics: Diagnostics
+    ): ReferenceMap<Expression.FunctionCall.Argument, Function.Parameter> {
```
Comment (tomekb234) [inline app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt:13]: A type alias `ArgumentResolutionResult` may be convenient.

File: app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt (hunk 2 above)
Comment (woocash2) [reply app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt:41]: I matched it to `NameResolutionErrors`. I think I'll convert this to singular as well.

File: app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt (hunk 3 above)
Comment (woocash2) [reply app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt:9]: Ok, then we can't have info like "modifies but doesn't read", but I think it's not necessary. ReadOnly and ReadWrite should be enough for our future usecases.

**Pull request merged**